### ISSUE TITLE
What is the reason behind converting the `cookie.maxAge` prop to an `Expires` value in the actual cookie?

### ISSUE THREAD

**Issue by bttger** [fastify/session#176]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the issue has not already been raised

### Issue

`getCookieOpts()` always converts the `maxAge` property to the `expires` property by creating a new Date and adding the `maxAge` to the current server's time. The returned object omits the `maxAge` property then. I believe this creates issues if the client's time is out of sync and goes against the expected behavior.

What is the rationale behind this design choice? To be honest, I hope the answer is not just "because express does it the same way" 😅

**Comment by bttger** [comment]
Oh, and I again stumbled upon this because I had the following issue: https://github.com/TimDaub/better-sqlite3-session-store/pull/13

**Comment by climba03003** [comment]
> I believe this creates issues if the client's time is out of sync and goes against the expected behavior.

If the client time is out of sync. `Max-Age` and `Expires` has no different, both will goes against the expected behavior.

> What is the rationale behind this design choice?

I didn't create the plugin in beginning. But for me, I would prefer `Expires` more than `Max-Age`.
The main reason is that it is more easy to see when it should be expired. For `Max-Age`, we need to calculate a bit before knowing the expire date-time.

**Comment by bttger** [comment]
> If the client time is out of sync. Max-Age and Expires has no different, both will goes against the expected behavior.

No, the `Max-Age` is a relative time while `Expires` is absolute. For the `Max-Age` it doesn't matter if the client's time is out of sync and I believe that this is also why it always takes precedence over the `Expires` value (according to [RFC 6265](https://www.rfc-editor.org/rfc/rfc6265#section-4.1.2.2))

> For Max-Age, we need to calculate a bit before knowing the expire date-time.

I am not sure if I understand what you'd need to calculate. When setting/touching the session, you would only instantiate a new `Date` like the current `getCookieOpts()` [does it already](https://github.com/fastify/session/blob/master/lib/getCookieOpts.js#L12) to be able to set it in the DB and check if a session is still valid later on. When setting/touching the actual cookie you would only pass the `cookie.maxAge` value without any calculation (except for maybe dividing by 1000 because for whatever reason express-session and fastify-session use milliseconds). What is there that you would need to calculate?

**Comment by climba03003** [comment]
> No, the Max-Age is a relative time

It is relative to the creation time, when the time of computer is out-of-sync. It will still calculated from the out-of-sync time.
How can it determine the time elapsed without referring to the computer time.

**Comment by bttger** [comment]
> It is relative to the creation time, when the time of computer is out-of-sync. It will still calculated from the out-of-sync time. How can it determine the time elapsed without referring to the computer time?

Yes, right. The elapsed time can only be reliably calculated assuming the client stays out of sync. But that is not what I want to discuss about. What I am asking is, why the `cookie.maxAge` is not applied to the cookie but converted to an `Expires` value instead. This makes no sense to me and definitely goes against the developers expactation.

**Comment by climba03003** [comment]
I believe the RFC note in below is a valid reason.

>       NOTE: Some existing user agents do not support the Max-Age attribute.  User agents that do not support the Max-Age attribute ignore the attribute.

**Comment by climba03003** [comment]
Please be aware that this package is made to be compatible to `express-session`. The decision on implementation is most likely affected by or inherit from `express-session`.

`@fastify/cookie` do allow you to set `Max-Age` if you wish.

Is that answer your question?

**Comment by Uzlopak** [comment]
Well for me expires makes much more sense than maxAge. E.g. you can easily debug and see when it will expire, while with maxAge you somehow have to know when the cookie was created and calculate it. Also expires is set to the same value as in the session store.

I dont see any benefit to use maxAge

**Comment by bttger** [comment]
> I believe the RFC note in below is a valid reason.
> 
> > ```
> >   NOTE: Some existing user agents do not support the Max-Age attribute.  User agents that do not support the Max-Age attribute ignore the attribute.
> > ```
> 
> Discussion in [`Express`](https://github.com/expressjs/session/issues/421) also mention about it.

The RFC note is from 2011 and, according to the link shared in the express discussion, Internet Explorer was the only browser not supporting it until version 8 which was in 2009. In all previous projects I have used Max-Age and never had issues.

**Comment by Uzlopak** [comment]
But if you think, that maxAge is much more awesome, you can provide a PR which adds the desired functionality.

**Comment by bttger** [comment]
> Please be aware that this package is made to be compatible to `express-session`. The decision on implementation is most likely affected by or inherit from `express-session`.
> 
> `@fastify/cookie` do allow you to set `Max-Age` if you wish.
> 
> Is that answer your question?

I'm happy for having fastify-session and not having to implement it myself, so I'll stick with it. Thanks for your quick answers.

**Comment by climba03003** [comment]
> But if you think, that maxAge is much more awesome, you can provide a PR which adds the desired functionality.

Unless `express-session` changed, I would be keep it for now.
This package is used to be similar or inherit the behavior of `express-session`.

**Comment by bttger** [comment]
> But if you think, that maxAge is much more awesome, you can provide a PR which adds the desired functionality.

Thank you, but I'll try to address the issue I have with the maintainer of the session store. I have already implemented a fix but we were discussing whether it makes sense to merge the PR.

**Comment by Uzlopak** [comment]
The sessionstore does not decide if maxAge or Expires is sent

**Comment by bttger** [comment]
> The sessionstore does not decide if maxAge or Expires is sent

Yes but it reads the `session.expires` and `session.maxAge` values to set the correct expiration time. And in case of the session store I am using, it only reads the `maxAge` value.

**Comment by TimDaub** [comment]
> I believe this creates issues if the client's time is out of sync and goes against the expected behavior.

I have a personal bias against expires and using it in my web apps as reasoning about time zones and a respective machine's configuration is dramatically more difficult in dynamic deployment environments compared to "just" setting a "maxAge" in seconds. "maxAge" is relative to the machine whereas expires requires knowledge about a machines timezone configuration. I personally would always favor maxAge over expires unless there is a very specific reason to prefer time zones (i never had that case).

> `getCookieOpts()` always converts the `maxAge` property to the `expires` property by creating a new Date and adding the `maxAge` to the current server's time.

This could be problematic, and an incompatibility issue as express/session is preferring maxAge over expires, see my comment where I point it out: https://github.com/TimDaub/better-sqlite3-session-store/pull/13#issuecomment-1277977207 I believe this is why @bttger opened the issue.


> No, the `Max-Age` is a relative time while `Expires` is absolute. For the `Max-Age` it doesn't matter if the client's time is out of sync and I believe that this is also why it always takes precedence over the `Expires` value (according to [RFC 6265](https://www.rfc-editor.org/rfc/rfc6265#section-4.1.2.2))

Nice. This makes a lot of sense to me considering that I've both used expires and maxAge in production and I found maxAge easier to work with in production where environments change. I quite from the RFC:

> If a cookie has both the Max-Age and the Expires attribute, the Max-
   Age attribute has precedence and controls the expiration date of the
   cookie.

> It is relative to the creation time, when the time of computer is out-of-sync. It will still calculated from the out-of-sync time.

IMO out-of-sync computers are irrelevant. Everything will be wrong and we can't mitigate so I'd suggest to ignore this case.

> Please be aware that this package is made to be compatible to `express-session`

see express/session readme: 

> Note The expires option should not be set directly; instead only use the maxAge option.

**Comment by TimDaub** [comment]
> don't know what you means by `expires` will have time-zone issue. It can contain `time-zone` information and it is not necessarily to provide a the same `time-zone` that match the computer. The time is computed by `server` and it provides the correct `time-zone` to the `client`.

Yeah so in production the problem will arise that e.g. you have a distributed team that assumes their local computer's time zone and then they e.g. say: "in one hour a cookie or cache should invalidate" and then they implement some local timezone math to set the expires header. But then later those devs may not be in control over where their application is hosted. E.g. they may be located in EU but then the app runs in Australia. Or the next day it runs in Bogota, and so they have to implement a expires calculation that considers the server's time zone too.

> I don't know what you means and see the readme? This plugin is actually inherit the same behavior as `express-session`.

From express/session
<img width="860" alt="Screenshot 2022-10-19 at 11 40 10" src="https://user-images.githubusercontent.com/2758453/196656062-c9c55987-b7fb-49ac-8f37-c006459cbd0f.png">

**Comment by Uzlopak** [comment]
A PR is welcome

**Comment by climba03003** [comment]
> Yeah so in production the problem will arise that

No really, hard to explain.

> From express/session

Again, the behavior is inherit from `express-session`.
You are using a public API note to argue with an internal transformation behavior same as `express-session`.

**Comment by climba03003** [comment]
> A PR is welcome

I personally recommend people to reach `express-session` first. 
Then we follow up later.

**Comment by TimDaub** [comment]
> Again, the behavior is inherit from `express-session`.
> You are using a public API note to argue with an internal transformation behavior same as `express-session`.

OK fair, I've been operating under the assumption that @bttger has a point but I have looked into express/session and fastify/session and the behaviors seem similar. I'm not sure what is @bttger's problem.

**Comment by Uzlopak** [comment]
Well we can always deviate from the express way by putting our deviation behind a flag.

**Comment by climba03003** [comment]
Maybe we do not align because of the missing `getter`?

https://github.com/expressjs/session/blob/1e3fc39a3fc3810772dfa00a7535a710042c42aa/session/cookie.js#L83-L108

**Comment by climba03003** [comment]
@TimDaub 
Would it be better to use `originalMaxAge` for the store? 
When I look through the source-code, it align the same value for both `maxAge` and `expires`.

Updated when using `expires`.
https://github.com/expressjs/session/blob/1e3fc39a3fc3810772dfa00a7535a710042c42aa/session/cookie.js#L60-L63

Updated when using `maxAge` indirectly.
https://github.com/expressjs/session/blob/1e3fc39a3fc3810772dfa00a7535a710042c42aa/session/cookie.js#L83-L95

It always exist from the output.
https://github.com/expressjs/session/blob/1e3fc39a3fc3810772dfa00a7535a710042c42aa/session/cookie.js#L119

**Comment by bttger** [comment]
> > Again, the behavior is inherit from `express-session`.
> > You are using a public API note to argue with an internal transformation behavior same as `express-session`.
> 
> OK fair, I've been operating under the assumption that @bttger has a point but I have looked into express/session and fastify/session and the behaviors seem similar. I'm not sure what is @bttger's problem.

Yes, the behavior seems similar, also for me. But I am unable to set the maxAge value in fastify-session so that your session store picks it up correctly. It always uses the default value (1 day). So, I set the maxAge value in fastify-session's options, it converts it to the expires value (plus it deletes the maxAge property from the options and replaces it with originalMaxAge), and then your session store only checks for the maxAge value in the `cookie.options`.

**Comment by bttger** [comment]
> Would it be better to use originalMaxAge for the store?

At first glance this looks like a good idea. But I am wondering, why all the other session stores ([redis-connect](https://github.com/tj/connect-redis/) for example) don't read the originalMaxAge or even the maxAge. They only read the expires value.

**Comment by jsumners** [comment]
> Yeah so in production the problem will arise that e.g. you have a distributed team that assumes their local computer's time zone and then they e.g. say: "in one hour a cookie or cache should invalidate" and then they implement some local timezone math to set the expires header. But then later those devs may not be in control over where their application is hosted. E.g. they may be located in EU but then the app runs in Australia. Or the next day it runs in Bogota, and so they have to implement a expires calculation that considers the server's time zone too.

That is not an issue this plugin should solve. Said team should understand that multiple time zones exist and write code accordingly. Use [luxon](https://github.com/moment/luxon) to make it easier.

According to https://www.rfc-editor.org/rfc/rfc6265#section-5.1.1 point 6, times should be in UTC anyway:

> Let the parsed-cookie-date be the date whose day-of-month, month,
       year, hour, minute, and second (in UTC) are the day-of-month-
       value, the month-value, the year-value, the hour-value, the
       minute-value, and the second-value, respectively.

**Comment by bttger** [comment]
> Which means only use expires is enough for a session-store.

But for some reason the value hasn't been transformed/converted yet when the session store reads it. I'll do some more analysis later. I don't want to shift this discussion now to the specific issue with the session store.

**Comment by bttger** [comment]
> But for some reason the value hasn't been transformed/converted yet when the session store reads it.

@climba03003 I saw your PR and I believe it fixes my issue. @TimDaub I adapted the tests in the better-sqlite3-session-store repository to use the actual express and fastify session `Cookie` class instead of plain javascript objects. So, hopefully we can merge your PR soon and close the PR I made in @TimDaub's repository.

I still believe, considering browsers have progressed in the last ten years, Max-Age is a better practice nowadays. But thanks for the discussion here and thanks for the PR that hopefully fixes it.
